cartographer/mapping/internal/2d/pose_graph/constraint_builder_2d_test.cc

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/mapping/internal/2d/pose_graph/constraint_builder_2d.h"
+
+#include <functional>
+
+#include "cartographer/common/internal/testing/thread_pool_for_testing.h"
+#include "cartographer/mapping/2d/submap_2d.h"
+#include "cartographer/mapping/internal/pose_graph/constraint_builder.h"
+#include "cartographer/mapping/internal/testing/test_helpers.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace cartographer {
+namespace mapping {
+namespace pose_graph {
+namespace {
+
+pose_graph::proto::ConstraintBuilderOptions GenerateConstraintBuilderOptions() {
+  const std::string kConstraintBuilderLua = R"text(
+    include "pose_graph.lua"
+    POSE_GRAPH.constraint_builder.sampling_ratio = 1
+    POSE_GRAPH.constraint_builder.min_score = 0
+    POSE_GRAPH.constraint_builder.global_localization_min_score = 0
+    return POSE_GRAPH.constraint_builder)text";
+  auto constraint_builder_parameters =
+      test::ResolveLuaParameters(kConstraintBuilderLua);
+  return CreateConstraintBuilderOptions(constraint_builder_parameters.get());
+}
+
+class MockCallback {
+ public:
+  MOCK_METHOD1(Run, void(const ConstraintBuilder2D::Result&));
+};
+
+TEST(ConstraintBuilder2DTest, CallsBack) {
+  common::testing::ThreadPoolForTesting thread_pool;
+  ConstraintBuilder2D constraint_builder(GenerateConstraintBuilderOptions(),
+                                         &thread_pool);
+  MockCallback mock;
```

Comment:
Given that you always instantiate a ThreadPool, ConstraintBuilder2D and MockCallback I would add a tes fixture. Also I think we will be adding more test cases to this test to ensure the new task based concurrency model is correct :)

Reply:
Done (making it TEST_F)